### PR TITLE
🎉 New Source: Fastbill [python cdk]

### PR DESCRIPTION
**Pull request by jhammarstedt**
## What
Adding Fastbill as a source

## How
Python CDK

## Recommended reading order
1. `x.java`
2. `y.python`

## 🚨 User Impact 🚨
Are there any breaking changes? What is the end result perceived by the user? If yes, please merge this PR with the 🚨🚨 emoji so changelog authors can further highlight this if needed.

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [x] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [x] Documentation updated
    - [x] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [x] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [x] `docs/integrations/README.md`
    - [x] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

<details><summary><strong>Connector Generator</strong></summary>

- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed

</details>

## Tests

<details><summary><strong>Unit</strong></summary>

*Put your unit tests output here.*

</details>

<details><summary><strong>Integration</strong></summary>

*Put your integration tests output here.*

</details>

<details><summary><strong>Acceptance</strong></summary>

*Put your acceptance tests output here.*

</details>

### CODE REVIEW

**Comment by marcosmarxm** [comment]
Thanks for your contribution @jhammarstedt! Please request my review when the code is ready to review. If you have any questions feel free to ask them in our Slack channel #hacktoberfest-2022. You can get a link to access Slack here: https://airbyte.com/connector-contest

**Comment by marcosmarxm** [comment]
@jhammarstedt can you sign the CLA?

**Comment by jhammarstedt** [comment]
@marcosmarxm ready for review!  Seems I can't tag you

**Review by marcosmarxm — CHANGES_REQUESTED** [review]
Hello @jhammarstedt, Marcos from Airbyte here 👋 . We received more than 25 new contributions along the weekend. One is yours 🎉  thank so much for! Our team is limited and maybe the review process can take longer than expected. As described in the Airbyte's Hacktoberfest your contribution was submitted before November 2nd and it is eligible to win the prize. The review process will validate other requirements. I ask to you patience until someone from the team review it. 

Because I reviewed some contributions for Hacktoberfest so far I saw some common patterns you can check in advance:
* [ ] Make sure you have added connector documentation to `/docs/integrations/`
* [ ] Remove the file `catalog` from `/integration_tests`
* [ ] Edit the `sample_config.json` inside `/integration_tests`
* [ ] For the `configured_catalog` you can use only `json_schema: {}`
* [ ] Add title to all properties in the `spec.yaml`
* [ ] Make sure the `documentationUrl` in the `spec.yaml` redirect to Airbyte's future connector page, eg: connector Airtable the `documentationUrl: https://docs.airbyte.com/integrations/sources/airtable`
* [ ] Review now new line at EOF (end-of-file) for all files.

If possible send to me a DM in Slack with the tests credentials, this process will make easier to us run integration tests and publish your connector. If you only has production keys, make sure to create a bootstrap.md explaining how to get the keys.

**Comment by jhammarstedt** [comment]
Hi, think I solved all the common patterns now, will send you my sandbox credentials as well.  
* [X]  Make sure you have added connector documentation to `/docs/integrations/` 
 * [X]  Remove the file `catalog` from `/integration_tests`
* [X]  Edit the `sample_config.json` inside `/integration_tests`
* [X]  For the `configured_catalog` you can use only `json_schema: {}`
* [X]  Add title to all properties in the `spec.yaml`
* [X]  Make sure the `documentationUrl` in the `spec.yaml` redirect to Airbyte's future connector page, eg: connector Airtable the `documentationUrl: https://docs.airbyte.com/integrations/sources/airtable`

* [X] If possible send to me a DM in Slack with the tests credentials, this process will make easier to us run integration tests and publish your connector. If you only has production keys, make sure to create a bootstrap.md explaining how to get the keys.

File: airbyte-config/init/src/main/resources/seed/source_definitions.yaml [hunk 1]
```diff
@@ -349,6 +349,13 @@
   documentationUrl: https://docs.airbyte.com/integrations/sources/faker
   sourceType: api
   releaseStage: alpha
+- name: Faker
```
Comment (marcosmarxm) [inline airbyte-config/init/src/main/resources/seed/source_definitions.yaml:352]: Please correct the name of the connector.

File: airbyte-integrations/connectors/source-fastbill/source_fastbill/schemas/TODO.md [hunk 2]
```diff
@@ -0,0 +1,25 @@
+# TODO: Define your stream schemas
+Your connector must describe the schema of each stream it can output using [JSONSchema](https://json-schema.org). 
+
+The simplest way to do this is to describe the schema of your streams using one `.json` file per stream. You can also dynamically generate the schema of your stream in code, or you can combine both approaches: start with a `.json` file and dynamically add properties to it.
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-fastbill/source_fastbill/schemas/TODO.md:4]: Remove this file.

File: airbyte-integrations/connectors/source-fastbill/source_fastbill/spec.yaml [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+  properties:
+    username:
+      title: Username
+      type: string
+      description: Username for Fastbill account
+    api_key:
+      title: API Key
+      type: string
+      description: Fastbill API key
+      airbyte_secret: true
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-fastbill/source_fastbill/spec.yaml:18]: Missing EOL.

File: airbyte-integrations/connectors/source-fastbill/source_fastbill/source.py [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+        super().__init__(*args, **kwargs)
+        self._username = username
+        self._api_key = api_key
+
+    API_OFFSET_LIMIT = 100
+
+    @property
+    def http_method(self) -> str:
+        return "POST"
+    url_base = " https://my.fastbill.com/api/1.0/api.php"
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-fastbill/source_fastbill/source.py:28]: Please check the code formatting.

File: airbyte-integrations/connectors/source-fastbill/source_fastbill/source.py [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+
+        response = response.json()
+        offset = response["REQUEST"]["OFFSET"] if response["REQUEST"]["OFFSET"] >= 0 else None
+        if offset is None:
+            response_request = response["REQUEST"]["OFFSET"]
+            raise Exception(f"No valid offset value found:{response_request}")
+
+        if len(response["RESPONSE"]["ARTICLES"]) == self.API_OFFSET_LIMIT:
+            return req_body(offset + self.API_OFFSET_LIMIT)
+        return None
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-fastbill/source_fastbill/source.py:188]: You repeating a lot of code, please create function for those.

**Review by marcosmarxm — CHANGES_REQUESTED** [review]
I made a review, there a lot of refactor (code duplication can be simplified)

File: airbyte-integrations/connectors/source-fastbill/source_fastbill/source.py (hunk 5 above)
Comment (jhammarstedt) [reply airbyte-integrations/connectors/source-fastbill/source_fastbill/source.py:188]: Added a helper.py where I store these, I can move them back to the source if it' preferred.

File: airbyte-config/init/src/main/resources/seed/source_definitions.yaml (hunk 1 above)
Comment (jhammarstedt) [reply airbyte-config/init/src/main/resources/seed/source_definitions.yaml]: Done

File: airbyte-integrations/connectors/source-fastbill/source_fastbill/source.py (hunk 4 above)
Comment (jhammarstedt) [reply airbyte-integrations/connectors/source-fastbill/source_fastbill/source.py]: Moved it to the top of the class instead

File: airbyte-integrations/connectors/source-fastbill/source_fastbill/spec.yaml (hunk 3 above)
Comment (jhammarstedt) [reply airbyte-integrations/connectors/source-fastbill/source_fastbill/spec.yaml]: Added

**Comment by jhammarstedt** [comment]
Hi @marcosmarxm, I fixed all the requested changes. Have a look and let me know if there's something else!

**Comment by jhammarstedt** [comment]
Hey @marcosmarxm, anything else I need to fix or is it ready to merge? :)

File: airbyte-integrations/connectors/source-fastbill/source_fastbill/source.py [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+    ) -> Optional[Union[Mapping, str]]:
+        return get_request_body_json(next_page_token, endpoint="customer")
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        return get_next_page_token(response=response, response_key="CUSTOMERS",
+                                   API_OFFSET_LIMIT=self.API_OFFSET_LIMIT,
+                                   endpoint="customer")
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        yield from response.json().get("RESPONSE", {}).get("CUSTOMERS", [])
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-fastbill/source_fastbill/source.py:147]: You can simplify again.
Create a data field, similar to `primary_key`.


```python
class Customers:
    endpoint = "CUSTOMERS"
    ...
    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
        yield from response.json().get("RESPONSE", {}).get(self.endpoint, [])
```

Comment (marcosmarxm) [reply airbyte-integrations/connectors/source-fastbill/source_fastbill/source.py:147]: You can use the same variable to improve the next_page_token function too.

**Review by marcosmarxm — CHANGES_REQUESTED** [review]
Some considerations about the code.

**Comment by jhammarstedt** [comment]
@marcosmarxm Thanks. fixed your suggested changes!

File: airbyte-integrations/connectors/source-fastbill/source_fastbill/source.py [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+    ) -> Optional[Union[Mapping, str]]:
+        return get_request_body_json(next_page_token, endpoint="article")
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        return get_next_page_token(response=response, response_key=self.data,
+                                   API_OFFSET_LIMIT=self.API_OFFSET_LIMIT,
+                                   endpoint="article")
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        yield from response.json().get("RESPONSE", {}).get(self.data, [])
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-fastbill/source_fastbill/source.py:107]: You can move this to parent class (`FastbillStream`) @jhammarstedt

Comment (marcosmarxm) [reply airbyte-integrations/connectors/source-fastbill/source_fastbill/source.py:107]: both functions next_page_token and parse_response.

**Review by marcosmarxm — CHANGES_REQUESTED** [review]
Last comment and it is ready to go :)

**Comment by jhammarstedt** [comment]
@marcosmarxm Done! You were right, it looked so clean just squashing all code together now as you suggested :)

**Comment by marcosmarxm** [comment]
Hello! I'm going to be out of the office this Friday and won't be able to review your contribution again today, I return next Monday. So far, most contributions look solid and are almost done to be approved. As said in [Chris' comment](https://airbytehq.slack.com/archives/C0450PCSDB2/p1667430678866629) all contributions made before 2-November are eligible to receive the prize and have 2 weeks to merge the contributions. But I ensure next week we're going to have your contribution merged. If you have questions about the implementation you can send them in `#hacktoberfest-2022` in Airbyte's Slack.

Sorry the inconvenience and see you again next week, thank you so much for your contribution!

**Review by marcosmarxm — APPROVED** [review]

**Pull request merged**